File: internal/handlers/nodes.go
// Node-inspector, topology proxy, and topology-editor handlers. Each handler
// fetches or mutates one newtron endpoint and returns the decoded "data" field
// as JSON.
//
// newtcon mirrors newtron's path geometry: every network-scoped resource lives
// under /api/networks/{netID}/... — netID is positional, extracted via
// r.PathValue("netID") in each handler. URL families:
//   - /api/networks/{netID}/topology
//   - /api/networks/{netID}/topology/nodes (write endpoints)
//   - /api/networks/{netID}/topology/links (write endpoints)
//   - /api/networks/{netID}/nodes/{device}/...
//
// All reads and writes are proxied through internal/newtronc/nodes.go; no
// newtron package is imported here (CLAUDE.md §newtron API Consumption Rule).
package handlers

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"strings"

	"github.com/aldrin-isaac/newtcon/internal/newtronc"
	"github.com/aldrin-isaac/newtcon/internal/types"
)

// NodesDeps is the dependency set for topology and node-inspector handlers.
type NodesDeps struct {
	Client        *newtronc.Client
	CorrelationID func(ctx context.Context) string
}

// RegisterNodesRoutes registers the topology, topology-editor, per-device read,
// and interface binding endpoints.
//
// Read endpoints:
//
//	GET  /api/networks/{netID}/topology                                — full topology
//	GET  /api/networks/{netID}/nodes/{device}/info                     — device overview
//	GET  /api/networks/{netID}/nodes/{device}/health                   — health check
//	GET  /api/networks/{netID}/nodes/{device}/interfaces               — interface list
//	GET  /api/networks/{netID}/nodes/{device}/interfaces/{name}        — interface detail
//	GET  /api/networks/{netID}/nodes/{device}/interfaces/{name}/binding — bound service
//	GET  /api/networks/{netID}/nodes/{device}/vlans                    — VLANs
//	GET  /api/networks/{netID}/nodes/{device}/vrfs                     — VRFs
//	GET  /api/networks/{netID}/nodes/{device}/acls                     — ACLs
//	GET  /api/networks/{netID}/nodes/{device}/lags                     — LAGs
//	GET  /api/networks/{netID}/nodes/{device}/neighbors                — neighbors
//	GET  /api/networks/{netID}/nodes/{device}/bgp/status               — BGP status
//	GET  /api/networks/{netID}/nodes/{device}/evpn/status              — EVPN status
//	GET  /api/networks/{netID}/nodes/{device}/configdb                 — full CONFIG_DB snapshot
//	GET  /api/networks/{netID}/nodes/{device}/configdb/{table}         — table keys
//	GET  /api/networks/{netID}/nodes/{device}/configdb/{table}/{key}   — entry value
//
// Topology write endpoints (operator-domain names):
//
//	POST   /api/networks/{netID}/topology/nodes                            — add device
//	PUT    /api/networks/{netID}/topology/nodes/{name}                     — update device
//	DELETE /api/networks/{netID}/topology/nodes/{name}                     — remove device
//	POST   /api/networks/{netID}/topology/links                            — add link
//	DELETE /api/networks/{netID}/topology/links/{device}/{interface}       — remove link
//
// Interface binding endpoints:
//
//	POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/bind-service    — bind service
//	POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/unbind-service  — unbind service
//	POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/refresh-service — refresh service
func RegisterNodesRoutes(mux *http.ServeMux, deps NodesDeps) {
	cid := deps.CorrelationID
	if cid == nil {
		cid = func(ctx context.Context) string { return "" }
	}
	c := deps.Client

	// proxyNode is the shared helper: calls fetchFn, writes the raw JSON payload
	// on success, or the standard error envelope on failure.
	proxyNode := func(
		w http.ResponseWriter,
		r *http.Request,
		fetchFn func(ctx context.Context) (json.RawMessage, error),
		endpoint string,
	) {
		ctx := r.Context()
		payload, err := fetchFn(ctx)
		if err != nil {
			writeUpstreamError(w, cid(ctx), err, endpoint, nil)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write(payload)
	}

	// Topology.
	mux.Handle("GET /api/networks/{netID}/topology", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.Topology(ctx, netID)
		}, "/api/networks/"+netID+"/topology")
	}))

	// Node overview.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/info", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeInfo(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/info")
	}))

	// Node health.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/health", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeHealth(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/health")
	}))

	// Interface list.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/interfaces", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeInterfaces(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/interfaces")
	}))

	// Interface detail. The interface name in the URL uses %2F for slashes
	// (e.g., Ethernet0%2F1 → Ethernet0/1), matching newtron's interfaceName()
	// normalisation. We replicate that normalisation here.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/interfaces/{name}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		name := normalizeIfaceName(r.PathValue("name"))
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeInterface(ctx, netID, device, name)
		}, "/api/networks/"+netID+"/nodes/"+device+"/interfaces/"+name)
	}))

	// Interface service binding.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/interfaces/{name}/binding", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		name := normalizeIfaceName(r.PathValue("name"))
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeInterfaceBinding(ctx, netID, device, name)
		}, "/api/networks/"+netID+"/nodes/"+device+"/interfaces/"+name+"/binding")
	}))

	// VLANs.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/vlans", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeVLANs(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/vlans")
	}))

	// VRFs.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/vrfs", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeVRFs(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/vrfs")
	}))

	// ACLs.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/acls", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeACLs(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/acls")
	}))

	// LAGs.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/lags", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeLAGs(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/lags")
	}))

	// Neighbors.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/neighbors", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeNeighbors(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/neighbors")
	}))

	// BGP status.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/bgp/status", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeBGPStatus(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/bgp/status")
	}))

	// EVPN status.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/evpn/status", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeEVPNStatus(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/evpn/status")
	}))

	// CONFIG_DB snapshot.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/configdb", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeConfigDB(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/configdb")
	}))

	// CONFIG_DB table keys.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/configdb/{table}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		table := r.PathValue("table")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeConfigDBTable(ctx, netID, device, table)
		}, "/api/networks/"+netID+"/nodes/"+device+"/configdb/"+table)
	}))

	// CONFIG_DB entry. The key may contain "/" characters; we capture the full
	// remainder after /{table}/ using a wildcard pattern.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/configdb/{table}/{key...}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		table := r.PathValue("table")
		key := r.PathValue("key")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeConfigDBEntry(ctx, netID, device, table, key)
		}, "/api/networks/"+netID+"/nodes/"+device+"/configdb/"+table+"/"+key)
	}))

	// Drift: comparison of intent vs CONFIG_DB reality. Newtron returns the
	// per-table differences; rendering surfaces these to the operator.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/drift", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeDrift(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/drift")
	}))

	// Intent projection: current logical state derived from intents.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/projection", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeProjection(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/projection")
	}))

	// Intent tree: the structured intent record graph.
	mux.Handle("GET /api/networks/{netID}/nodes/{device}/intent-tree", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeIntentTree(ctx, netID, device)
		}, "/api/networks/"+netID+"/nodes/"+device+"/intent-tree")
	}))

	// Reconcile: dry_run=true returns the drift preview; dry_run=false executes
	// the corrective intent push. mode=topology drives full reconcile.
	mux.Handle("POST /api/networks/{netID}/nodes/{device}/reconcile", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		dryRun := r.URL.Query().Get("dry_run") == "true"
		mode := r.URL.Query().Get("mode")
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeReconcile(ctx, netID, device, dryRun, mode)
		}, "/api/networks/"+netID+"/nodes/"+device+"/reconcile")
	}))

	// Generic RPC: forward any node-level newtron action by subpath.
	// Query string is forwarded verbatim so callers can pass newtron options
	// like ?mode=topology and ?execute=false.
	mux.Handle("POST /api/networks/{netID}/nodes/{device}/rpc/{subpath...}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		subpath := r.PathValue("subpath")
		rawQuery := r.URL.RawQuery
		body, _ := io.ReadAll(r.Body)
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.NodeRPC(ctx, netID, device, subpath, rawQuery, body)
		}, "/api/networks/"+netID+"/nodes/"+device+"/rpc/"+subpath)
	}))

	// Generic per-interface RPC.
	mux.Handle("POST /api/networks/{netID}/nodes/{device}/interfaces/{iface}/rpc/{subpath...}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		iface := normalizeIfaceName(r.PathValue("iface"))
		subpath := r.PathValue("subpath")
		rawQuery := r.URL.RawQuery
		body, _ := io.ReadAll(r.Body)
		proxyNode(w, r, func(ctx context.Context) (json.RawMessage, error) {
			return c.InterfaceRPC(ctx, netID, device, iface, subpath, rawQuery, body)
		}, "/api/networks/"+netID+"/nodes/"+device+"/interfaces/"+iface+"/rpc/"+subpath)
	}))

	// ============================================================================
	// Topology write endpoints
	// ============================================================================

	// proxyNodeWrite is the shared helper for write (POST/PUT/DELETE) routes that
	// forward to newtron and return the data field, mapping errors to the standard
	// newtcon envelope.
	proxyNodeWrite := func(
		w http.ResponseWriter,
		r *http.Request,
		statusOnSuccess int,
		writeFn func(ctx context.Context) (json.RawMessage, error),
		endpoint string,
	) {
		ctx := r.Context()
		payload, err := writeFn(ctx)
		if err != nil {
			writeUpstreamError(w, cid(ctx), err, endpoint, nil)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(statusOnSuccess)
		_, _ = w.Write(payload)
	}

	// readBody decodes a JSON request body into the provided pointer.
	// Returns false and writes the error envelope when the body is invalid.
	readBody := func(w http.ResponseWriter, r *http.Request, v any) bool {
		body, err := io.ReadAll(r.Body)
		if err != nil {
			types.WriteError(w, http.StatusBadRequest, types.KindValidationFailure,
				"reading request body: "+err.Error(), nil)
			return false
		}
		if err := json.Unmarshal(body, v); err != nil {
			types.WriteError(w, http.StatusBadRequest, types.KindValidationFailure,
				"invalid JSON: "+err.Error(), nil)
			return false
		}
		return true
	}

	// POST /api/networks/{netID}/topology/nodes — add a device to the topology.
	// Forwards body {name, device} to newtron's create-node verb.
	mux.Handle("POST /api/networks/{netID}/topology/nodes", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		var body any
		if !readBody(w, r, &body) {
			return
		}
		proxyNodeWrite(w, r, http.StatusCreated, func(ctx context.Context) (json.RawMessage, error) {
			return c.CreateTopologyDevice(ctx, netID, body)
		}, "POST /api/networks/"+netID+"/topology/nodes")
	}))

	// PUT /api/networks/{netID}/topology/nodes/{name} — update a device entry (full replacement).
	mux.Handle("PUT /api/networks/{netID}/topology/nodes/{name}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		name := r.PathValue("name")
		var body any
		if !readBody(w, r, &body) {
			return
		}
		proxyNodeWrite(w, r, http.StatusOK, func(ctx context.Context) (json.RawMessage, error) {
			return c.UpdateTopologyDevice(ctx, netID, name, body)
		}, "PUT /api/networks/"+netID+"/topology/nodes/"+name)
	}))

	// DELETE /api/networks/{netID}/topology/nodes/{name} — remove a device from the topology.
	// Query param ?force=true cascade-deletes referring links.
	mux.Handle("DELETE /api/networks/{netID}/topology/nodes/{name}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		name := r.PathValue("name")
		force := r.URL.Query().Get("force") == "true"
		proxyNodeWrite(w, r, http.StatusOK, func(ctx context.Context) (json.RawMessage, error) {
			return c.DeleteTopologyDevice(ctx, netID, name, force)
		}, "DELETE /api/networks/"+netID+"/topology/nodes/"+name)
	}))

	// POST /api/networks/{netID}/topology/links — add a link between two interfaces.
	// Forwards body {a, z} to newtron's create-link verb.
	mux.Handle("POST /api/networks/{netID}/topology/links", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		var body any
		if !readBody(w, r, &body) {
			return
		}
		proxyNodeWrite(w, r, http.StatusCreated, func(ctx context.Context) (json.RawMessage, error) {
			return c.CreateTopologyLink(ctx, netID, body)
		}, "POST /api/networks/"+netID+"/topology/links")
	}))

	// DELETE /api/networks/{netID}/topology/links/{device}/{interface} — remove the link that
	// includes the given endpoint. Newtron resolves the full link from a single endpoint.
	mux.Handle("DELETE /api/networks/{netID}/topology/links/{device}/{interface}", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		iface := normalizeIfaceName(r.PathValue("interface"))
		proxyNodeWrite(w, r, http.StatusOK, func(ctx context.Context) (json.RawMessage, error) {
			return c.DeleteTopologyLink(ctx, netID, device, iface)
		}, "DELETE /api/networks/"+netID+"/topology/links/"+device+"/"+iface)
	}))

	// ============================================================================
	// Interface binding endpoints
	// ============================================================================

	// POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/bind-service
	// Body: { service: string, ip_address?: string, vlan?: int, peer_as?: int, params?: object }
	mux.Handle("POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/bind-service", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		name := normalizeIfaceName(r.PathValue("name"))
		var body any
		if !readBody(w, r, &body) {
			return
		}
		proxyNodeWrite(w, r, http.StatusOK, func(ctx context.Context) (json.RawMessage, error) {
			return c.ApplyService(ctx, netID, device, name, body)
		}, "POST /api/networks/"+netID+"/nodes/"+device+"/interfaces/"+name+"/bind-service")
	}))

	// POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/unbind-service
	mux.Handle("POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/unbind-service", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		name := normalizeIfaceName(r.PathValue("name"))
		proxyNodeWrite(w, r, http.StatusOK, func(ctx context.Context) (json.RawMessage, error) {
			return c.RemoveService(ctx, netID, device, name)
		}, "POST /api/networks/"+netID+"/nodes/"+device+"/interfaces/"+name+"/unbind-service")
	}))

	// POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/refresh-service
	mux.Handle("POST /api/networks/{netID}/nodes/{device}/interfaces/{name}/refresh-service", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := r.PathValue("netID")
		device := r.PathValue("device")
		name := normalizeIfaceName(r.PathValue("name"))
		proxyNodeWrite(w, r, http.StatusOK, func(ctx context.Context) (json.RawMessage, error) {
			return c.RefreshService(ctx, netID, device, name)
		}, "POST /api/networks/"+netID+"/nodes/"+device+"/interfaces/"+name+"/refresh-service")
	}))
}

// normalizeIfaceName converts %2F sequences to "/" in interface names from
// URL path values. Matches newtron's interfaceName() normalisation.
func normalizeIfaceName(name string) string {
	return strings.ReplaceAll(name, "%2F", "/")
}

File: web/src/api/newtcon/nodes.ts
// Typed client for newtcon-server's topology and node-inspector endpoints.
//
// All calls pass through newtcon-server which proxies to newtron verbatim.
// The data field is returned as-is (unknown) so callers can render with the
// recursive renderValue helper in app.ts without coupling to a concrete type.
//
// Every function targets the operator's active network by default; pass
// `network` to target a specific network (cross-engine workflows).

import { apiFetch, apiSend } from "./_transport.js";
import { apiPath } from "../../api-path.js";

function pathFor(suffix: string, network?: string): string {
  return network ? apiPath.network(network, suffix) : apiPath(suffix);
}

// fetchNodeRaw delegates to apiFetch with no-cache. Kept as a thin local
// alias so per-endpoint functions read consistently.
function fetchNodeRaw(url: string): Promise<unknown> {
  return apiFetch(url, { cache: "no-store" });
}

// fetchTopology returns the full topology payload from GET /api/networks/{netID}/topology.
export async function fetchTopology(network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor("topology", network));
}

// fetchNodeInfo returns device overview.
export async function fetchNodeInfo(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/info`, network));
}

// fetchNodeHealth returns health data.
export async function fetchNodeHealth(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/health`, network));
}

// fetchNodeInterfaces returns the interface list.
export async function fetchNodeInterfaces(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/interfaces`, network));
}

// fetchNodeInterface returns detail for one interface.
export async function fetchNodeInterface(device: string, ifaceName: string, network?: string): Promise<unknown> {
  // Encode "/" in interface names as %2F so the path segment is unambiguous.
  const encodedName = ifaceName.replace(/\//g, "%2F");
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/interfaces/${encodedName}`, network));
}

// fetchNodeInterfaceBinding returns the service binding for one interface.
export async function fetchNodeInterfaceBinding(device: string, ifaceName: string, network?: string): Promise<unknown> {
  const encodedName = ifaceName.replace(/\//g, "%2F");
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/interfaces/${encodedName}/binding`, network));
}

// fetchNodeVLANs returns VLAN status.
export async function fetchNodeVLANs(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/vlans`, network));
}

// fetchNodeVRFs returns VRF list.
export async function fetchNodeVRFs(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/vrfs`, network));
}

// fetchNodeACLs returns ACL list.
export async function fetchNodeACLs(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/acls`, network));
}

// fetchNodeLAGs returns LAG list.
export async function fetchNodeLAGs(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/lags`, network));
}

// fetchNodeNeighbors returns neighbors.
export async function fetchNodeNeighbors(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/neighbors`, network));
}

// fetchNodeBGPStatus returns BGP status.
export async function fetchNodeBGPStatus(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/bgp/status`, network));
}

// fetchNodeEVPNStatus returns EVPN status.
export async function fetchNodeEVPNStatus(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/evpn/status`, network));
}

// fetchNodeConfigDB returns the full CONFIG_DB snapshot.
export async function fetchNodeConfigDB(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/configdb`, network));
}

// fetchNodeConfigDBTable returns the key list for one CONFIG_DB table.
export async function fetchNodeConfigDBTable(device: string, table: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/configdb/${encodeURIComponent(table)}`, network));
}

// fetchNodeConfigDBEntry returns one CONFIG_DB entry.
export async function fetchNodeConfigDBEntry(
  device: string,
  table: string,
  key: string,
  network?: string,
): Promise<unknown> {
  return fetchNodeRaw(pathFor(
    `nodes/${encodeURIComponent(device)}/configdb/${encodeURIComponent(table)}/${encodeURIComponent(key)}`,
    network,
  ));
}

// fetchNodeDrift returns intent-vs-CONFIG_DB drift for a device.
export async function fetchNodeDrift(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/drift`, network));
}

// fetchNodeProjection returns the projected intent state.
export async function fetchNodeProjection(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/projection`, network));
}

// fetchNodeIntentTree returns the structured intent record graph.
export async function fetchNodeIntentTree(device: string, network?: string): Promise<unknown> {
  return fetchNodeRaw(pathFor(`nodes/${encodeURIComponent(device)}/intent-tree`, network));
}

// postNodeReconcile triggers reconcile: dry_run=true returns drift preview;
// dry_run=false executes corrective intent push.
export async function postNodeReconcile(
  device: string,
  opts: { dryRun: boolean; mode?: string } = { dryRun: true },
  network?: string,
): Promise<unknown> {
  const params = new URLSearchParams();
  if (opts.dryRun) params.set("dry_run", "true");
  if (opts.mode) params.set("mode", opts.mode);
  const qs = params.toString();
  const base = pathFor(`nodes/${encodeURIComponent(device)}/reconcile`, network);
  return apiSend(qs ? `${base}?${qs}` : base, "POST");
}

// ============================================================================
// Write helpers (topology editor + interface binding)
// ============================================================================

// nodeWrite is a thin wrapper for write methods. Delegates to apiSend.
function nodeWrite(
  url: string,
  method: "POST" | "PUT" | "DELETE",
  body?: unknown,
): Promise<unknown> {
  return apiSend(url, method, body);
}

// ---- Topology editor --------------------------------------------------------

// postTopologyDevice adds a device to the topology.
// body: { name: string, device: { steps?: ..., ports?: ... } }
export async function postTopologyDevice(body: Record<string, unknown>, network?: string): Promise<unknown> {
  return nodeWrite(pathFor("topology/nodes", network), "POST", body);
}

// putTopologyDevice replaces a device entry (full replacement).
// body: TopologyDevice — { steps?: ..., ports?: ... }
export async function putTopologyDevice(name: string, body: Record<string, unknown>, network?: string): Promise<unknown> {
  return nodeWrite(pathFor(`topology/nodes/${encodeURIComponent(name)}`, network), "PUT", body);
}

// deleteTopologyDevice removes a device from the topology.
// force=true cascade-deletes referring links.
export async function deleteTopologyDevice(name: string, force = false, network?: string): Promise<unknown> {
  const base = pathFor(`topology/nodes/${encodeURIComponent(name)}`, network);
  const url = force ? `${base}?force=true` : base;
  return nodeWrite(url, "DELETE");
}

// postTopologyLink adds a link between two interfaces.
// body: { a: "device:interface", z: "device:interface" }
export async function postTopologyLink(body: { a: string; z: string }, network?: string): Promise<unknown> {
  return nodeWrite(pathFor("topology/links", network), "POST", body);
}

// deleteTopologyLink removes the link that includes the given endpoint.
export async function deleteTopologyLink(device: string, ifaceName: string, network?: string): Promise<unknown> {
  const encodedIface = ifaceName.replace(/\//g, "%2F");
  return nodeWrite(
    pathFor(`topology/links/${encodeURIComponent(device)}/${encodedIface}`, network),
    "DELETE",
  );
}

// ---- Interface service binding ----------------------------------------------

// postBindService binds a service to an interface.
// body: { service: string, ip_address?: string, vlan?: number, peer_as?: number, params?: object }
export async function postBindService(
  device: string,
  ifaceName: string,
  body: Record<string, unknown>,
  network?: string,
): Promise<unknown> {
  const encodedIface = ifaceName.replace(/\//g, "%2F");
  return nodeWrite(
    pathFor(`nodes/${encodeURIComponent(device)}/interfaces/${encodedIface}/bind-service`, network),
    "POST",
    body,
  );
}

// postUnbindService removes the service binding from an interface (no body).
export async function postUnbindService(device: string, ifaceName: string, network?: string): Promise<unknown> {
  const encodedIface = ifaceName.replace(/\//g, "%2F");
  return nodeWrite(
    pathFor(`nodes/${encodeURIComponent(device)}/interfaces/${encodedIface}/unbind-service`, network),
    "POST",
  );
}

// postRefreshService re-applies the bound service on an interface (no body).
export async function postRefreshService(device: string, ifaceName: string, network?: string): Promise<unknown> {
  const encodedIface = ifaceName.replace(/\//g, "%2F");
  return nodeWrite(
    pathFor(`nodes/${encodeURIComponent(device)}/interfaces/${encodedIface}/refresh-service`, network),
    "POST",
  );
}

// postNodeRPC POSTs to any node-level newtron action via the generic
// /api/networks/{netID}/nodes/{device}/rpc/{subpath} proxy. body may be
// null/empty for actions that take no params.
export async function postNodeRPC(
  device: string,
  subpath: string,
  body: Record<string, unknown> | null = null,
  network?: string,
): Promise<unknown> {
  const url = pathFor(`nodes/${encodeURIComponent(device)}/rpc/${subpath}`, network);
  return apiSend(url, "POST", body && Object.keys(body).length > 0 ? body : undefined);
}

// postInterfaceRPC POSTs to any per-interface newtron action.
export async function postInterfaceRPC(
  device: string,
  iface: string,
  subpath: string,
  body: Record<string, unknown> | null = null,
  network?: string,
): Promise<unknown> {
  const ifaceEnc = iface.replace(/\//g, "%2F");
  const url = pathFor(`nodes/${encodeURIComponent(device)}/interfaces/${encodeURIComponent(ifaceEnc)}/rpc/${subpath}`, network);
  return apiSend(url, "POST", body && Object.keys(body).length > 0 ? body : undefined);
}

// postProjectionDiff POSTs a list of {url, params} ops to newtron's
// per-device intent/projection-diff endpoint via the newtcon-server
// proxy. Powers the per-device projection in the apply-preview modal
// (slice #171.B). Operations apply in-memory only on newtron and the
// substrate state is restored before the response returns.
export async function postProjectionDiff(
  device: string,
  ops: Array<{ url: string; params: Record<string, unknown> }>,
  network?: string,
): Promise<unknown> {
  const url = pathFor(`nodes/${encodeURIComponent(device)}/projection-diff`, network);
  return apiSend(url, "POST", { operations: ops });
}
